### PR TITLE
Retry cluster pipeline read errors

### LINKED ISSUE #930: cluster client doesn't retry on pipeline read timeouts
Here is an example:

```go
package main

import (
	"fmt"
	"time"

	"github.com/go-redis/redis"
)

func main() {
	fmt.Println("start")
	options := &redis.ClusterOptions{
		Addrs:              []string{"0.0.0.0:7000"},
		MaxRedirects:       8,
		MaxRetries:         8,
		MinRetryBackoff:    1 * time.Second,
		MaxRetryBackoff:    1 * time.Minute,
		DialTimeout:        100 * time.Millisecond,
		ReadTimeout:        100 * time.Millisecond,
		WriteTimeout:       100 * time.Millisecond,
	}
	c := redis.NewClusterClient(options)
	fmt.Printf("%+v\n", c.Options())
	defer c.Close()
	fmt.Println("ping", c.Ping())
	options.ReadTimeout = time.Millisecond
	pipe := c.Pipeline()
	for i := 0; i < 128; i++ {
		z := redis.Z{
			Score:  float64(0),
			Member: fmt.Sprintf("test%d", i),
		}
		pipe.ZAdd("z", z)
	}
	_, err := pipe.Exec()
	fmt.Println(err)
}
```

- `pipelineReadCmds` doesn't update the `failedCmds` map https://github.com/go-redis/redis/blob/master/cluster.go#L1352.
- `defaultProcessPipeline` doesn't retry because `failedCmds` is empty https://github.com/go-redis/redis/blob/master/cluster.go#L1275.

### PR DESCRIPTION
**Pull request by akrylysov**
Fixes #930.

### CODE REVIEW

File: cluster.go [hunk 1]
```diff
@@ -1366,9 +1366,12 @@ func (c *ClusterClient) pipelineReadCmds(
 			continue
 		}
 
-		return err
+		cmd.setErr(err)
```
Comment (vmihailenco) [inline cluster.go]: this is not needed - cmd already has an error after readReply

File: cluster.go [hunk 2]
```diff
@@ -1366,9 +1366,12 @@ func (c *ClusterClient) pipelineReadCmds(
...
+		failedCmds.mu.Lock()
+		failedCmds.m[node] = append(failedCmds.m[node], cmd)
+		failedCmds.mu.Unlock()
 	}
-	return nil
+	return cmdsFirstErr(cmds)
```
Comment (vmihailenco) [inline cluster.go]: If there is a `redis.Nil` it will be returned here. I think this is wrong.

File: cluster_test.go [hunk 3]
```diff
@@ -683,6 +683,40 @@ var _ = Describe("ClusterClient", func() {
...
+
+			startRetry := time.Now()
+			_, err = clientRetry.Pipelined(func(pipe redis.Pipeliner) error {
+				pipe.Ping()
+				return nil
+			})
+			Expect(err).To(HaveOccurred())
+			elapseRetry := time.Since(startRetry)
+
+			Expect(elapseRetry).To(BeNumerically(">", elapseNoRetry, 10*time.Millisecond))
```
Comment (vmihailenco) [inline cluster_test.go]: Well, this is not very reliable. Yes, clientRetry does more work than clientNoRetry but that is the only thing we test. No guarantees it does anything useful...

Comment (akrylysov) [reply cluster_test.go]: I followed the pattern used in the `should retry with backoff` test (https://github.com/go-redis/redis/blob/4b568cdf1adcfe5cab06ad6929ad5286b3db7138/redis_test.go#L158). Can you suggest a better way of testing it?

**Comment by akrylysov** [comment]
@vmihailenco feel free to close my PR and merge yours, thanks!

File: cluster_test.go (hunk 3 above)
Comment (vmihailenco) [reply cluster_test.go]: No, not really - sorry.

**Comment by vmihailenco** [comment]
I've merged https://github.com/go-redis/redis/pull/936/files - hopefully it will not cause any problems.

Thanks for this PR anyway. It was very helpful.

**Pull request closed without merging**